nixos/modules/services/backup/rustic.nix

```diff
@@ -0,0 +1,445 @@
+          # jq's `fromdate` builtins do not support sub-second accuracy (as of early 2025)
+          (
+            m:
+            ''| gsub("{{${m}}}"; $json.summary.${m} | sub(".[0-9]{0,9}Z"; "Z") | fromdateiso8601 | tostring)''
+          )
+          prometheusTimeMetrics
+        }
+      ' > "${f}/rustic-backup-$1.prom.next"
+      mv "${f}/rustic-backup-$1.prom.next" "${f}/rustic-backup-$1.prom"
+    '';
+
+  commonOptions = {
+    startAt = lib.mkOption {
+      type = with lib.types; either (listOf str) str;
+      description = ''
+        Time(s) at which to run this operation.
+
+        The format is documented in `man systemd.time`.
+      '';
+    };
+
+    useProfiles = lib.mkOption {
+      type = lib.types.listOf lib.types.str;
+      default = [ ];
+      description = "Config profiles to use";
+    };
+
+    extraArgs = lib.mkOption {
+      type = lib.types.listOf lib.types.str;
+      default = [ ];
+      description = "Extra command-line arguments to pass to the `rustic` call.";
+    };
+  };
+
+  commonBackupOptions = commonOptions // {
+    pipeJsonInto = lib.mkOption {
+      description = "Command that will be called on the JSON result of `rustic backup`.";
+      default = {
+        nothing = { };
+      };
+      type = lib.types.attrTag {
+        nothing = lib.mkOption {
+          description = "Do nothing with the result.";
+          type = lib.types.unspecified;
+        };
+
+        command = lib.mkOption {
+          description = ''
+            Run a command on the result.
+
+            It will be called with the JSON result on standard input, and an identifier of
+            the backup section as an argument. You can pass any other arguments eg. as tags,
+            which rustic will return in its JSON output.
+          '';
+          type = lib.types.str;
+        };
+
+        prometheus = lib.mkOption {
+          description = "Write the result as prometheus metrics.";
+          type = lib.types.submodule {
+            options.nodeExporterCollectorFolder = lib.mkOption {
+              type = lib.types.path;
+              description = ''
+                Folder that is exported by the node exporter, and in which the metrics should be placed.
+
+                This should be configured in the node exporter as `--collector.textfile.directory=`.
+              '';
+            };
+          };
+        };
+      };
+    };
+  };
+in
+{
+  options.services.rustic = {
+    enable = lib.mkEnableOption "rustic";
+    package = lib.mkPackageOption pkgs "rustic" { };
+
+    checkProfiles = lib.mkOption {
+      type = lib.types.bool;
+      default = true;
+      description = ''
+        If enabled, checks that the generated rustic profiles are valid.
+
+        Note that imports that cannot be resolved are accepted. This allows
+        the recommended setup of using an imported file that only root can read
+        to store the passwords.
+      '';
+    };
+
+    profiles = lib.mkOption {
```

Comment:
Is the idea that the operator can specify `profiles` and/or `backups`, with `profiles` being files the operator writes himself while `backups` is generated from a config?

Reply:
The idea is that `profiles` determines the configuration of `rustic`, whereas `backups` defines a specific backup target.

The split allows a user to eg. define a `profile` that's not used for any automated backup, to trigger on-demand backups with `sudo rustic -P my-manual-backup`.

OTOH, having `backups` without `profiles` would likely be pretty useless.